src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -100,18 +100,18 @@ public static async ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(
             // Use local pipe to avoid blocking the current thread on networking IO.
             var localPipe = new Pipe(PipeOptionsWithUnlimitedWriterBuffer);
 
-            Exception? exception = null;
-
             // start a task on a thread pool thread copying from the RPC pipe to a local pipe:
             var copyTask = Task.Run(async () =>
             {
+                Exception? exception = null;
+
                 try
                 {
                     await pipeReader.CopyToAsync(localPipe.Writer, cancellationToken).ConfigureAwait(false);
                 }
-                catch (Exception e)
+                catch (Exception e) when ((exception = e) == null) // capture, but don't catch the exception
```

Comment:
You were propagating it before as far as I can tell. You haven't even altered the timing of that. 
Is it that you're now propagating a *different* (inner?) exception than you were before? Why would you want to do that? Propagating the outer exception together with its inner exceptions has always been good enough for us.
Oh... might this actually lead to propagating exceptions that were *caught and handled* as well, as if they were fatal? That doesn't sound like a good idea, if in fact that is the effect it has.

Reply:
The only change this PR makes is that instead of storing the exception into a local and then rethrowing it when we see EndOfStreamException, we let the copyTask fail (the exception is not caught) and then await of the copyTask in the finally block will throw it.

Reply:
Previously, it could potentially happen that the exception is thrown in CopyToAsync, then EndOfStreamException is thrown in ReadData and handled before the local variable is assigned to.

Reply:
> Previously, it could potentially happen that the exception is thrown in CopyToAsync, then EndOfStreamException is thrown in ReadData and handled before the local variable is assigned to.

I don't see how this could happen. `localPipe.Writer` is only closed _after_ `exception` has a value.

Reply:
Assuming the pipe is operating correctly. I think it's better to not assume that.